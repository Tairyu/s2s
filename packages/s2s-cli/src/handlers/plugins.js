// @flow
import path from 'path'
import fs from 'fs'
import KeyLocker from 'key-locker'
import {
  getOutputPath,
  writeFileSync,
  toErrorStack,
  log,
  relativeFromCwd,
  resolveInputPath,
} from '../utils'
import type { Path, AfterHook, Plugin, EventType, HanlderFunc } from '../types'
import runHooks from '../hooks/run'
import { formatText } from '../reporters/'
import babelHandler from './babel-handler'

type Opts = {
  eventPath: Path,
  plugin: Plugin,
  hooks: AfterHook[],
}

export function handlePlugin(
  handler: HanlderFunc,
  { eventPath, plugin, hooks = [] }: Opts
) {
  const filename = resolveInputPath(plugin.input, eventPath)
  const content = fs.readFileSync(filename, 'utf8')

  const code = handler(content, { eventPath, plugin, filename })

  if (!code && code === '') {
    return
  }

  const result = runHooks(
    resolveInputPath(plugin.input, eventPath),
    code,
    hooks
  )

  const outputPath = plugin.output
    ? getOutputPath(
        plugin.output.replace('[name]', path.parse(eventPath).name),
        eventPath
      )
    : eventPath

  writeFileSync(outputPath, result)
  log(formatText('S2S', relativeFromCwd(eventPath), outputPath))
}

const lock = new KeyLocker()

export default function handlePlugins(
  eventPath: Path,
  eventType: EventType,
  plugins: Plugin[] = [],
  hooks: AfterHook[] = []
) {
  if (lock.has(eventPath)) {
    return
  }

  for (const plugin of plugins) {
    try {
      if (!plugin.test.test(eventPath)) {
        continue // eslint-disable-line
      }

      if (plugin.only && !plugin.only.includes(eventType)) {
        continue // eslint-disable-line
      }

      lock.add(eventPath)

      const handler = plugin.handler ? plugin.handler : babelHandler

      if (handler) {
        handlePlugin(handler, { eventPath, plugin, hooks })
      }
    } catch (err) {
      console.error(toErrorStack(err))
    }
  }
}
