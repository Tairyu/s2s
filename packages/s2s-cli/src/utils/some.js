// @flow
import micromatch from 'micromatch'

export default function(list: string | string[], patterns: string | string[]) {
  list = [].concat(list)
  for (let i = 0; i < list.length; i++) {
    if (micromatch(list[i], patterns).length === 1) {
      return true
    }
  }
  return false
}
