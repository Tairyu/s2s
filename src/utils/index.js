// @flow
import path from 'path'
import fs from 'fs'
import { transformFileSync } from 'babel-core'

type Path = string
type Filename = string
type Code = string

export function log(msg: string) {
  console.log(msg)
}

export function compile(
  filename: Filename,
  opts: Object
): { code?: Code, ignored: boolean } {
  try {
    return transformFileSync(filename, opts)
  } catch (err) {
    console.error(toErrorStack(err))
    return { ignored: true }
  }
}

export function toErrorStack(err: Object) {
  if (err._babel && err instanceof SyntaxError) {
    return `${err.name}: ${err.message}\n${err.codeFrame}`
  }
  return err.stack
}

export function getOutputPath(output: Path, input: Path): Path {
  if (path.isAbsolute(output)) {
    return output
  }

  return path.resolve(path.dirname(input), output)
}

export function writeFileSync(outputPath: Path, code: string) {
  fs.writeFileSync(outputPath, code, 'utf-8')
}

export function relativeFromCwd(input: Path) {
  return path.relative(process.cwd(), input)
}

export function getDirAndBaseName(testPath: string) {
  const dirname = path.dirname(testPath)
  const basename = path.basename(testPath)

  return {
    dirname,
    basename,
  }
}
