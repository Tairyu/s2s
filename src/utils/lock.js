// @flow

export default class Lock {
  _lock: Set<string>
  _timer: ?number

  constructor() {
    this._lock = new Set()
  }

  has(input: string) {
    return this._lock.has(input)
  }

  add(input: string) {
    this.clear()
    clearTimeout(this._timer)

    this._lock.add(input)

    this._timer = setTimeout(() => {
      this._lock.delete(input)
    }, 1000)
  }

  clear() {
    this._lock.clear()
  }
}
