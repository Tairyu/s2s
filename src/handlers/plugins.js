// @flow
import { compile, getOutputPath, write, toErrorStack } from '../utils'
import type { Path, AfterHook, Plugin, PluginOpts } from '../types'
import runHooks from '../hooks/run'
import { log } from '../utils'
import lock from '../utils/lock'
import { formatText } from '../reporters/'

export function compileWithPlugin(input: Path, plugin: PluginOpts) {
  const { code } = compile(input, {
    babelrc: false,
    plugins: [plugin],
  })
  return code ? code.trim() : ''
}

export function handlePlugin(
  eventPath: Path,
  plugin: Plugin,
  hooks: AfterHook[] = []
) {
  if (!plugin.test.test(eventPath)) {
    return
  }
  lock.add(eventPath)

  const input = plugin.input ? plugin.input : eventPath

  const code = compileWithPlugin(input, plugin.plugin)

  if (!code && code === '') {
    return
  }

  const result = runHooks(input, code, hooks)

  const outputPath = plugin.output
    ? getOutputPath(plugin.output, eventPath)
    : eventPath

  write(outputPath, result)

  log(formatText('S2S', input, outputPath))
}

export default function handlePlugins(
  input: Path,
  plugins: Plugin[] = [],
  hooks: AfterHook[] = []
) {
  if (lock.has(input)) {
    lock.clear()
    return
  }

  for (const plugin of plugins) {
    try {
      handlePlugin(input, plugin, hooks)
    } catch (err) {
      console.error(toErrorStack(err))
    }
  }
}
