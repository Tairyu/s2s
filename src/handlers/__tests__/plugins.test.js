// @flow
import path from 'path'
import * as utils from '../../utils'
import lock from '../../utils/lock'
import * as plugins from '../plugins'
import _plugin from './helpers/identifer-reverse-plugin'

let errorSpy
let logSpy
let writeSpy

const fn = x => x

const getEventPath = () => path.resolve(__dirname, 'fixtures', 'a.js')

beforeEach(() => {
  errorSpy = jest.spyOn(console, 'error').mockImplementation(fn)
  logSpy = jest.spyOn(utils, 'log').mockImplementation(fn)
  writeSpy = jest.spyOn(utils, 'writeFileSync').mockImplementation(fn)
})

afterEach(() => {
  errorSpy.mockRestore()
  logSpy.mockRestore()
  writeSpy.mockRestore()
})

test('compileWithPlugin', () => {
  const result = plugins.compileWithPlugin(getEventPath(), _plugin)
  expect(result).toMatchSnapshot()
})

test('handlePlugin when eventPath not match', () => {
  const spyFn = jest.spyOn(plugins, 'compileWithPlugin')
  const plugin = { test: /hoge/, plugin: _plugin }
  plugins.handlePlugin(getEventPath(), plugin)
  expect(spyFn.mock.calls.length).toBe(0)
})

test('handlePlugin when eventPath match', () => {
  const plugin = { test: /a.js/, plugin: _plugin }
  plugins.handlePlugin(getEventPath(), plugin)
  expect(logSpy.mock.calls[0][0]).toMatch('S2S')
})

test('handlePlugin with input option', () => {
  const plugin = { test: /a.js/, plugin: _plugin, input: getEventPath() }
  plugins.handlePlugin(getEventPath(), plugin)
  expect(writeSpy.mock.calls[0][1]).toMatchSnapshot()
})

test('handlePlugin with output option', () => {
  const plugin = { test: /a.js/, plugin: _plugin, output: 'b.js' }
  plugins.handlePlugin(getEventPath(), plugin)
  expect(writeSpy.mock.calls[0][0]).toMatch('/fixtures/b.js')
})

test('handlePlugin write args', () => {
  const plugin = { test: /a.js/, plugin: _plugin }
  plugins.handlePlugin(getEventPath(), plugin)
  expect(writeSpy.mock.calls[0][0]).toBe(getEventPath())
  expect(writeSpy.mock.calls[0][1]).toMatchSnapshot()
})

test('handlePlugin when compileWithPlugin returns empty code', () => {
  const spy = jest.spyOn(utils, 'compile').mockReturnValue('')
  const plugin = { test: /a.js/, plugin: _plugin }
  plugins.handlePlugin(getEventPath(), plugin)
  expect(logSpy).not.toHaveBeenCalled()
  spy.mockRestore()
})

test('handlePlugins when locked', () => {
  lock.add(getEventPath())
  plugins.default(getEventPath(), [_plugin])
  expect(writeSpy).not.toHaveBeenCalled()
})

test('handlePlugins called write', () => {
  const opts = { test: /not-found.js/, plugin: _plugin }
  plugins.default(getEventPath(), [opts])
  expect(logSpy).not.toHaveBeenCalled()
})

test('handlePlugins handle error', () => {
  const plugin = { test: /a.js/, plugin: _plugin }
  logSpy.mockImplementation(() => {
    throw new Error('hello')
  })
  plugins.default(getEventPath(), [plugin])
  expect(errorSpy).toHaveBeenCalled()
})

test('handlePlugin when plugins = undefined', () => {
  plugins.default(getEventPath())
  expect(logSpy).not.toHaveBeenCalled()
})
