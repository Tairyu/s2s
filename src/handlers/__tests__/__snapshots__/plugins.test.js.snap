// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileWithPlugin 1`] = `
"// @flow
const olleh = 'hello';"
`;

exports[`handlePlugin with input option 1`] = `
"// @flow
const olleh = 'hello';"
`;

exports[`handlePlugin write args 1`] = `
"// @flow
const olleh = 'hello';"
`;
